keps/sig-api-machinery/customresources-pruning-and-defaulting.md

```diff
@@ -0,0 +1,486 @@
+>
+>```json
+>{"anyOf": [{"properties": {"a": {}}}, {"properties": {"b": {}}}]}
+>```
+>
+>Then a JSON object `{"a":1, "b": 2, "c": 3}` is pruned to `{"a":1, "b":2}` with the given semantics. Note that also `{"a":1}`, `{"b":2}` would be natural pruning results if we defined "specified" in a different way.
+
+#### Example 3
+
+> Assume the OpenAPI schema
+>
+>```json
+>{
+>   "properties": {
+>       "a": {}, 
+>       "b": {"type": "string", "properties": {"x": {}}}
+>   }, 
+>   "anyOf": [
+>	    { "properties": {"c":{}} }, 
+>       { "properties": {"d":{"not":{}}}}
+>   ]
+>}
+>```
+>
+> Then `a`, `b`, `b.x`, `c`, `d` are specified, `b.y` and `e` are not specified.
+
+**Question:** is it natural semantics to assume `d` to be specified and not to prune it? Note that there is no object with `d` that validates against `"d":{not:{}}`. But due to the `anyOf` there are objects which will validate against the complete schema.
+
+#### Pruning Options
+
+Motivated by the examples, we have different options to define the pruning semantics:
+1. Use the described semantics of `specified`.
+2. Only consider `properties` fields in the schema which actually successfully validate a given object (then `d` would be pruned in example 3).
+3. Only consider `properties` fields outside of `anyOf`, `allOf`, `oneOf`, `not` during pruning.
+4. Only consider `properties` fields outside of `anyOf`, `allOf`, `oneOf`, `not` during pruning, but enforce that every `properties` key inside `anyOf`, `allOf`, `oneOf`, `not` also appears outside all of those.
+
+From these options:
+1. leads to fields being kept from pruning although they only appear in branches of the OpenAPI validation schema which do not contribute to validation of an object.
+2. is ambiguous if you have multiple branches of `anyOf` validating. Should we drop the fields of the first or the second branch?
+3. leads to surprising pruning of fields that the user forgot to specify outside of propositional logic of `anyOf`, `allOf`, `oneOf`, `not`.
+4. forces the user to re-specify all those properties outside of `anyOf`, `allOf`, `oneOf`, `not` which appear inside of them. The outcome will match that of 1, with the difference that it makes the skeleton explicit.
+
+**Here we propose not to follow 2 and 3 due to ambiguity of 2 and the danger of user mistakes of 3. Both 1 and 4 lead to the same pruning behaviour and only differ in whether the user has to re-specify property keys or whether they are automatically derived.**
+
+### Restrictions on Defaults
+
+The OpenAPI validation schema allows to define defaults via the `default` field. For example:
+
+```json
+{"properties": {"a": {"default": {"x": 42}}}}
+```
+
+An empty object `{}` would be defaulted to `{"a":{"x":42}}`
+
+
+Obviously, the default values should validate. But OpenAPI validation schemata do not enforce that. Also one default in a branch of `anyOf`, `allOf`, `oneOf` might force another validation run onto another branch of these quantors.
```

Comment:
**David Eads:** this doesn't bother me

**Stefan Schimanski:** We do a second round of validation in the registry validation step. So we catch them.

Reply:
We propose below to only allow defaults outside of `anyOf`, `allOf`, ... So we don't have to decide about values and which branches they fulfill for them, making application of defaults very simple as well.

Reply:
>  it will converge as the number of branches is finite and defaulting is idem-potent. But it is a strange behaviour nevertheless.

I'm pretty sure that this is turing complete or close to it if you can engineer branches that drop fields. Which is more or less what this KEP is proposing.

I really think we need to look hard at the the schema and come up with some rules that prevent people from saying things that they can't say with go types. I have some thoughts on how to do this, but this KEP isn't the place to hash it out. We should NOT be able to get turing-complete behavior by iterated default application.